### DISCUSSION TITLE
Vehicle Routing Problem with Time Windows  change example

### DISCUSSION THREAD

**Discussion by ra-esmith** [google/or-tools#3385] [Routing (and legacy CP) questions] [answered]
Hello,

How can I have one vehicle and different time windows?

I took the example https://developers.google.com/optimization/routing/vrptw  provided and made a couple changes.  First I changed the time windows to 0 to 100 - thinking this would let any of the times work.  That worked fine.

But then I changed to one vehicle, instead of 4.  And it did not work.  Just hanged and never found a solution.

Why not?  Why would it not be ok with 1 vehicle with these changes.

Trying to understand this example of time windows in more detail.

Thanks!
Evan

**Comment by Mizux** [comment, marked as answer]
Did you also increase the vehicle maximum time ?
aka
```py
routing.AddDimension(
    transit_callback_index,
    30,  # allow waiting time
    30,  # maximum time per vehicle
    False,  # Don't force start cumul to zero.
    time)
```
here you should change `30` by at least `100` since all your TW are 0-100 according to your message...

**Comment by ra-esmith** [comment]
great, will try, thanks!

**Comment by ra-esmith** [comment]
YES!  ok so having increased the allowed waiting time to 100 and the max time per vehicle to 100
runs just fine

great, thanks !

**Comment by ra-esmith** [comment]
Great, I made the windows more complicated by adding
        time_dimension.CumulVar(index).RemoveInterval(10, 20)
        To my example
        
Checking the source code
https://github.com/google/or-tools/blob/2cb85b4eead4c38e1c54b48044f92087cf165bce/ortools/constraint_solver/constraint_solver.h#L4053-L4061

To see the API

The result is
Route for vehicle 0:
0 Time(0,0) -> 9 Time(2,2) -> 5 Time(4,4) -> 8 Time(6,6) -> 6 Time(8,8) -> 2 Time(11,11) -> 10 Time(15,15) -> 16 Time(19,19) -> 14 Time(21,21) -> 13 Time(25,25) -> 12 Time(27,27) -> 11 Time(28,28) -> 15 Time(31,31) -> 3 Time(37,37) -> 4 Time(38,38) -> 1 Time(40,40) -> 7 Time(44,44) -> 0 Time(46,46)



which is intereseting ... in that I would have thought the time for say location 10 from the example would not have the time 15,15 ... as I blocked the times from 10 to 20?

**Comment by ra-esmith** [comment]
the code change to the example is
```py
    # Add time window constraints for each vehicle start node.
    depot_idx = data['depot']
    for vehicle_id in range(data['num_vehicles']):
        index = routing.Start(vehicle_id)
        time_dimension.CumulVar(index).SetRange(
            data['time_windows'][depot_idx][0],
            data['time_windows'][depot_idx][1])
        time_dimension.CumulVar(index).RemoveInterval(10, 20)
```

**Reply by Mizux** [reply]
here you remove interval [10,20] for vehicle start node aka node 0 since vehicle start at Time(0,0) it's OK ;)
if you want ot remove it for node 10
```python
index = manager.NodeToIndex(10)
time_dimension.CumulVar(index).RemoveInterval(10, 20)
```
note: by default range is [0, dimension vehicle max capacity] for all nodes

**Comment by ra-esmith** [comment]
AH, silly me.

Got it
Updated example
```py
    # Add time window constraints for each location except depot.
    for location_idx, time_window in enumerate(data['time_windows']):
        if location_idx == data['depot']:
            continue
        index = manager.NodeToIndex(location_idx)
        time_dimension.CumulVar(index).SetRange(time_window[0], time_window[1])
        time_dimension.CumulVar(index).RemoveInterval(10, 20)        
    # Add time window constraints for each vehicle start node.
    depot_idx = data['depot']
    for vehicle_id in range(data['num_vehicles']):
        index = routing.Start(vehicle_id)
        time_dimension.CumulVar(index).SetRange(
            data['time_windows'][depot_idx][0],
            data['time_windows'][depot_idx][1])
```
Leaving the depot alone and updating all the indexes

result is thus
```sh
Objective: 46
Route for vehicle 0:
0 Time(0,0) -> 9 Time(2,3) -> 5 Time(4,5) -> 8 Time(6,7) -> 6 Time(8,9) -> 2 Time(21,21) -> 10 Time(25,25) -> 16 Time(29,29) -> 14 Time(31,31) -> 13 Time(35,35) -> 12 Time(37,37) -> 11 Time(38,38) -> 15 Time(41,41) -> 3 Time(47,47) -> 4 Time(48,48) -> 1 Time(50,50) -> 7 Time(54,54) -> 0 Time(56,56)
Time of the route: 56min
```
and I have as expected no times between 10 and 20!

cool.

**Comment by ra-esmith** [comment]
Question about how the time matrix could be "distances"
Lets say that I use lat/lng and convert the time matrix into distances from each location.  Easy enough, say in miles (but a big number as all over USA).

Do the units of the "time windows" need to change?  How does the "size" of the time window vs the units of my matrix relate here?  As far as can tell they are orthogonal .. and are not related?

Have not tried and example of that and am thinking of doing that next

**Reply by Mizux** [reply]
dimension are unit less or put differently are the unit you decide so supposing your matrix are numbers in miles so the intervals represent a miles intervals aka vehicle can visit the node if it's "miles counter" is not between 10 and 20 miles to use your previous example so your TimeWindows are actually DistanceWindows (which should be of little sense unless you want to model gaz/maintenance station ?)

note: SlackVar can be seen as "turning around" for doing extra miles for fun... so usually for distance dimension you want a SlackVar to be in range [0,0] ;)

**Comment by ra-esmith** [comment]
[example.txt](https://github.com/google/or-tools/files/9132490/example.txt)
Ok here I have added a list of locations (lat/lng) I convert them into a "time matrix" which is I guess ok as a distance matrix (which is what it is).  And kept the windows 0 to 100 for time. 

Yet no solution found.

What am I missing ...

**Reply by Mizux** [reply]
printing your distance matrix (aka `print(dst_matrix)`)
```py
[[0, 1070, 1218, 685, 1046, 3274, 3679, 1950, 2267, 1792],
 [1070, 0, 418, 520, 1359, 3749, 3878, 1079, 1380, 938],
 [1218, 418, 0, 859, 1734, 3463, 3519, 1382, 1653, 1261],
 [685, 520, 859, 0, 877, 3766, 4029, 1264, 1582, 1107],
 [1046, 1359, 1734, 877, 0, 4313, 4719, 1597, 1863, 1467],
 [3274, 3749, 3463, 3766, 4313, 0, 1022, 4823, 5111, 4686],
 [3679, 3878, 3519, 4029, 4719, 1022, 0, 4899, 5152, 4781],
 [1950, 1079, 1382, 1264, 1597, 4823, 4899, 0, 317, 157],
 [2267, 1380, 1653, 1582, 1863, 5111, 5152, 317, 0, 475],
 [1792, 938, 1261, 1107, 1467, 4686, 4781, 157, 475, 0]]
```
So with your ~~time~~ distance dimension still in range [0,100] vehicles will have serious difficulty to travel the few thousand unit your matrix contains ;)

ps: You should use our discord (chat badge in the readme)

**Comment by ra-esmith** [comment]
Hello,

Ok will look for the read me ... nto sure what a discord or chat badge is.  will fix.

Ok so the distance matrix and the "travel time" are related in that they must have the same scale?

I wanted to have "distance between locations".  And then visit locations say 2 per week.  But not sure how to do that then.  As clearly different scales

**Comment by ra-esmith** [comment]
I I am hoping to have
distances between locations
and then time windows in days
... so that is where I am heading

**Reply by Mizux** [reply]
Then you may need two dimensions:
* one Distance dimension and
* one Time dimension.
(and two matrix (dst matrix, time matrix) so 2 callbacks to create and register)

note: Actually you can have any number of dimension you want/need

**Comment by ra-esmith** [comment]
Ok in the original example we have
    transit_callback_index = routing.RegisterTransitCallback(time_callback)

    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)


This is time in the original, and I of course changed it to distance (as there are other examples like that).
This is not a "dimension" I think?  But is require I think.

Thus how does on specify the RegisterTransitCallback which is for me distance
but at the same time have a distance dimension and time dimension

And "time" here is not so much the time between locations.  But time visiting.  For example, I want to plan my vacation, and want to visit some major cities, chicago, new york, san fran, and austin.  And want to spend 1 day at each ... and what is the best path.



Thanks
Evan

**Comment by ra-esmith** [comment]
and I should add that while on vacation I know I can't visit say new york in august (as to hot) and can't go to chicago in winter (as to cold).

**Comment by ra-esmith** [comment]
So I have seen folks doing things like
```py
def distance_callback(from_index, to_index):
    """Returns the distance between the two nodes."""
    # Convert from routing variable Index to distance matrix NodeIndex.
    from_node = manager.IndexToNode(from_index)
    to_node = manager.IndexToNode(to_index)
    return data['distance_matrix'][from_node][to_node] + duration['to_index']
```    
Which is mixing the distance with the duration ... which sort of makes sense.  But only if they were in the same "units".
    
But hard to see how that would work out.

**Comment by ra-esmith** [comment]
So from the ORTools docs I am really thinking about a TSP problem.  With some time windows.
Not clear to me how that may change things here, but worth I think noting

**Comment by ra-esmith** [comment]
Hello,

No luck with badge and discord (I did create an account though :-)).  

Ok so taking
[example_tsp.txt](https://github.com/google/or-tools/files/9133608/example_tsp.txt)

https://developers.google.com/optimization/routing/tsp
We make the distances using lat/lng and a matrix

And now wish to I think add a time constraint so can spend "some time" at each location.

But expect it will conflict with the distances ...

**Reply by Mizux** [reply]
did you click on this link ? https://discord.gg/ENkQrdf

**Comment by ra-esmith** [comment]
[example_tsp.txt](https://github.com/google/or-tools/files/9133701/example_tsp.txt)
And as soon as I add a "time dimension" I get no solution, even though I gave 10000 as the time which is way more than the units of distance.

But how would I have distance in miles and time in days.  
Clearly missing something, but not sure yet what :-)  

Thanks !!!
Evan

**Comment by ra-esmith** [comment]
Referencing: https://acrogenesis.com/or-tools/documentation/user_manual/manual/tsp/first_tsptw_implementation.html



 Dimensions are quantities accumulated along the nodes in a routing solution and can be used to model time windows. Remember the formula from last section:

Total travel time to node j = Total travel time to i + Distance(i,j) + Service time at node i + Waiting time at i.

Which means to me that the "unites" of time and distance must somehow be able to relate.  And that is my question how?
If I have a year of 365 
and distance in miles
Then how do I combine them here, or anywhere when I have dimensions etc?

**Reply by Mizux** [reply]
again I don't understand why you want ot combine them, there are complementary in pseudo code you should have something like this with **two** dimension IMHO:

```py
## Distance constraint
distance_ matrix = create_distance_matrix_in_miles()

def distance_callback(from_index, to_index):
    """Returns the distance in miles between the two nodes."""
    # Convert from routing variable Index to distance matrix NodeIndex.
    from_node = manager.IndexToNode(from_index)
    to_node = manager.IndexToNode(to_index)
    return distance_matrix[from_node][to_node]

distance_callback_index = routing.RegisterTransitCallback(distance_callback)

# Add Distance constraint.
dimension_name = 'Distance'
routing.AddDimension(
  distance_callback_index,
  0,  # no slack
  10_000,  # vehicle maximum travel distance in miles
  True,  # start cumul to zero
  dimension_name)

distance_dimension = routing.GetDimensionOrDie(dimension_name)

## Time Constraints
duration_ matrix = create_duration_matrix_in_days()
duration_at_node = create_array_of_time_pass_at_each_location_in_days()

def time_callback(from_index, to_index):
    """Returns the time in days between the two nodes."""
    # Convert from routing variable Index to distance matrix NodeIndex.
    from_node = manager.IndexToNode(from_index)
    to_node = manager.IndexToNode(to_index)
    return duration_at_node[from_node] + duration_matrix[from_node][to_node]

time_callback_index = routing.RegisterTransitCallback(time_callback)

# Add Time constraint.
dimension_name = 'Time'
routing.AddDimension(
  time_callback_index,
  365,  # allow to wait between trip
  365,  # vehicle maximum travel duration in days aka visit every location within a year
  False,  # don't force start cumul to zero e.g. start our trip in February aka 32 not 0
  dimension_name)

time_dimension = routing.GetDimensionOrDie(dimension_name)

# Add time windows constraint
# e.g. location 3 (Chicago) remove interval [0-60] and [305,365]  ([Jan-Feb] and [Nov-Dec] as too cold))
index = manager.NodeToIndex(3)
time_dimension.CumulVar(index).RemoveInterval(0, 60) # [Jan-Feb]
time_dimension.CumulVar(index).RemoveInterval(304, 365) # [Nov-Dec]

# Objective
# Define cost of each arc reusing the distance callback.
# note: so here the objective function is distance based aka solver will try to minimize the total distance travelled
# and time is only use to limit the max (365) and to setup some time windows if you want.
routing.SetArcCostEvaluatorOfAllVehicles(distance_callback_index)

....
routing.solve(...)
```

**Reply by ra-esmith** [reply]
wow! great, will read and try out.  thanks!

**Comment by ra-esmith** [comment]
[vacation.txt](https://github.com/google/or-tools/files/9144591/vacation.txt)
When I run it seems to "work"
The route: (index, min time, max time)
[(0, 0, 0), (3, 3, 344), (4, 6, 347), (8, 9, 350), (7, 201, 352), (9, 203, 354), (1, 206, 357), (2, 208, 359), (5, 211, 362), (6, 214, 365)]

Two Questions ...
1) I would like to now extend this example
Plan vacation, visit cities at lat/lng - 3 per week, each on their own day
But not sure how I might go about that yet.


2) Also I am printing out the three tuple above (node, start, end)
The start and end are the min and max ... and I must admit that I am unsure what they mean.  I have a "remove" (block) on the time window from 10 to 200 on all the nodes - thus why there are no numbers 10 to 200.  But, for example (8, 9, 350).  Does this mean one can visit node 8 from 0 to 9 and 350 to 365?  But there are other nodes that do that ... or are each of these meaning I could visit always on the min or always on the max day?

**Reply by Mizux** [reply]
For 1. I need to think of it...

For 2. Start reading https://github.com/google/or-tools/issues/1051#issuecomment-461033132
- try to display the SlackVar (first you need to Add it to the assignment aka save it in the solution object)
ref: https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/samples/cvrptw.py#L190

ps:  you may be interested in https://github.com/Mizux/tsp_multiple_days/blob/feature/morning_nodes/40_Locations_3_days.py
pps: sorry 12pm here time to go to bed, wake at 5am to go to office ^^; will try to explain better tomorrow morning

**Comment by ra-esmith** [comment]
# Question 2 
GREAT!
Added duration, and now want to understand this output.

The current script ... such as it is :-)
[sample_vac.txt](https://github.com/google/or-tools/files/9149325/sample_vac.txt)


With the slack var my output is
Objective: 11967
  0 Time(0, 0) with Slack (0, 341)
  5 Time(3, 344) with Slack (0, 341)
  6 Time(6, 347) with Slack (0, 341)
  2 Time(9, 350) with Slack (0, 341)
  1 Time(201, 352) with Slack (0, 151)
  3 Time(204, 355) with Slack (0, 151)
  4 Time(207, 358) with Slack (0, 151)
  8 Time(210, 361) with Slack (0, 151)
  7 Time(212, 363) with Slack (0, 151)
  9 Time(214,365) with Slack (none as end)
None
Time of the route: 214 min
Total time of all routes: 214min

Question: how does one "read" this (what does it mean).
I am thinking ...

For example
8 Time(210, 361) with Slack (0, 151)
7 Time(212, 363) with Slack (0, 151)

Clearly one is at node 8 going to node 7 (locations)
The earliest one gets to node 8 is 210, or node 7 is 212 (in our day units)
Due to my durationintervalvar (of 3) we have 210 to 212 I think (always a jump of 2) - but not sure ...

Then one can "wait" at a node (aka the slack) of anything from 0 (which is above) or 151 days!  thus
210 + 151 being 361  - the max at the node.


Yet, how does one say an exact time?  Or perhaps one always gives a route in terms of min/max?  
Could I say node 8 at 215 and node 7 at 217 arrival day and be correct?   How would one not have a min/max in the end ... and have a fixed day one is arriving..


# Question 1: 
Plan vacation, visit cities at lat/lng - 3 per week, each on their own day
One could just change the duration_at_node to be 3 for each city, that is clear enough.  But how does one make the travel time duration a min of 4 days always.  I guess I could have a "pad" matrix - which ensures that travel time is 4 days and add that to the time dimension.  But not sure if this is how one goes about this.

# Question 3 
I found I had to add in 
    vehicle_id = 0        
    index = routing.Start(vehicle_id)
    time_dimension.CumulVar(index).SetRange(0, 0)
    routing.AddToAssignment(time_dimension.SlackVar(index))
Or else I got a crash ... but am not sure why I need this


# Question 4 
I don't have a distance dimension (no solution found when I had that), instead just registered transit callback for distance ... and am puzzled if I have set that up wrong.

**Comment by ra-esmith** [comment]
The things I wish to support for my vacation planner ... and not clear to me what rule/constraint I could write
* max number cities per week and max cities for month
* max distance between cities i visit
* max distance travel in a month

**Comment by ra-esmith** [comment]
Question: when running a TSP (with a set of locations) how might one find an optimal path for a "subset" of the locations - so perhaps not all of them?  So if perhaps I had 50 locations ... how might I choose the best path that is 20 long?  Not sure how one structures such a problem using ortools ...

Thanks as always
Evan

**Comment by azaharudue** [comment]
Hi, 
For a time window dependent VRP task I encountered a problem, 
The Assignment variable : `solution `  returns **null** whenever the vehicle number is less than 4. 
For testing simple case, 
I was trying to run the [example in documentation]( https://developers.google.com/optimization/routing/vrptw) too just like @ra-esmith mentioned [here](https://github.com/google/or-tools/discussions/3385?sort=top#discussion-4225465)
But in my case I didn't change anything other than just the vehicleNumber reducing  from 4 to 1. 
Could anyone please help me out in this issue ?
